joern-cli/frontends/jimple2cpg/src/main/scala/io/joern/jimple2cpg/Jimple2Cpg.scala

```diff
@@ -51,46 +52,77 @@ class Jimple2Cpg extends X2CpgFrontend[Config] {
+      Options.v().set_force_android_jar(framework)
+    } else {
+      Options.v().set_src_prec(Options.src_prec_apk_c_j)
+    }
+    Options.v().set_process_multiple_dex(true)
+  }
+
+  def sootLoadClass(inputDir: String): Unit = {
+    Options.v().set_process_dir(List(inputDir).asJava)
+    Options.v().set_src_prec(Options.src_prec_class)
+  }
+
+  def sootLoadJimple(inputDir: String): Unit = {
+    Options.v().set_process_dir(List(inputDir).asJava)
+    Options.v().set_src_prec(Options.src_prec_jimple)
+  }
+
+  def sootLoadJava(inputDir: String): Unit = {
+    // NOTE: Soot’s frontend for Java source files is outdated (only partially supports Java version up to 7) and not very robust.
+    Options.v().set_process_dir(List(inputDir).asJava)
+    Options.v().set_src_prec(Options.src_prec_java)
+  }
+
   def createCpg(config: Config): Try[Cpg] = {
     val ret = withNewEmptyCpg(config.outputPath, config: Config) { (cpg, config) =>
-      val rawSourceCodeFile = new JFile(config.inputPath)
-      val sourceTarget      = rawSourceCodeFile.toPath.toAbsolutePath.normalize.toString
-      val sourceCodeDir = if (rawSourceCodeFile.isDirectory) {
-        sourceTarget
-      } else {
-        Paths
-          .get(new JFile(sourceTarget).getParentFile.getAbsolutePath)
-          .normalize
-          .toString
-      }
+      val inputPath = new JFile(config.inputPath)
 
       configureSoot()
       new MetaDataPass(cpg, language, config.inputPath).createAndApply()
 
-      val sourceFileExtensions  = Set(".class", ".jimple")
-      val archiveFileExtensions = Set(".jar", ".war")
-      // Load source files and unpack archives if necessary
-      val sourceFileNames = if (sourceTarget == sourceCodeDir) {
-        // Load all source files in a directory
-        loadSourceFiles(sourceCodeDir, sourceFileExtensions, archiveFileExtensions)
+      if (inputPath.isDirectory()) {
+        // make sure classpath is configured correctly
+        Options.v().set_soot_classpath(ProgramHandlingUtil.getUnpackingDir.toString)
+        Options.v().set_prepend_classpath(true)
+        val sourceFileExtensions  = Set(".class", ".jimple")
+        val archiveFileExtensions = Set(".jar", ".war")
+        // Load source files and unpack archives if necessary
+        val sourceFileNames = loadSourceFiles(config.inputPath, sourceFileExtensions, archiveFileExtensions)
+        logger.info(s"Loading ${sourceFileNames.size} program files")
+        logger.debug(s"Source files are: $sourceFileNames")
+        loadClassesIntoSoot(sourceFileNames)
+        val astCreator = new AstCreationPass(sourceFileNames, cpg)
+        astCreator.createAndApply()
+        new TypeNodePass(astCreator.global.usedTypes.keys().asScala.toList, cpg)
+          .createAndApply()
       } else {
-        // Load single file that was specified
-        loadSourceFiles(sourceTarget, sourceFileExtensions, archiveFileExtensions)
+        val ext = config.inputPath.split("\\.").last
```

Comment:
I would recommend using the `lastOption` type here in case there is some weird input.